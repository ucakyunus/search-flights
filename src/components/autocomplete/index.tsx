import React, { forwardRef } from 'react';
import {
  Autocomplete as MUIAutocomplete,
  AutocompleteProps as MUIAutocompleteProps,
  InputAdornment,
  TextField,
} from '@mui/material';
import { ReturnLocation } from '@/types/location';

export interface AutocompleteProps
  extends Omit<MUIAutocompleteProps<ReturnLocation, false, false, false>, 'renderInput'> {
  startAdornment?: React.ReactNode;
  placeholder?: string;
}

export const Autocomplete = forwardRef(
  ({ placeholder, startAdornment, ...rest }: AutocompleteProps, forwardedRef) => {
    return (
      <MUIAutocomplete
        clearIcon={null}
        noOptionsText='Sonuç bulunamadı'
        sx={{ width: '100%', bgcolor: 'white' }}
        getOptionLabel={(option) => option.label}
        getOptionKey={(option) => option.value}
        renderInput={(params) => (
          <div style={{ height: '100%' }}>
            <TextField
              {...params}
              sx={{
                height: '100%',
                '.MuiOutlinedInput-root': {
                  height: '100%',
                },
              }}
              placeholder={placeholder}
              variant='outlined'
              inputRef={forwardedRef}
              slotProps={{
                input: {
                  ...params.InputProps,
                  startAdornment: (
                    <InputAdornment position='start' sx={{ color: 'primary.dark' }}>
                      {startAdornment}
                    </InputAdornment>
                  ),
                  sx: {
                    fontSize: '1.3rem',
                  },
                },
              }}
              fullWidth
            />
          </div>
        )}
        {...rest}
      />
    );
  },
);
