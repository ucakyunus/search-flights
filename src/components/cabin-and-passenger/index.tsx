import React, { useMemo, useState } from 'react';
import { IconButton, Typography } from '@mui/material';
import ManIcon from '@mui/icons-material/Man';
import AddIcon from '@mui/icons-material/Add';

import { CabinSelectorPopover } from '@/components/cabin-and-passenger/cabin-selector-popover';
import { FareCategories } from '@/constants';

export interface CabinAndPassengerProps {
  onPersonCount: (count: number) => void;
  onCabinType: (cabinType: FareCategories) => void;
  personCount: number;
  cabinType: FareCategories;
}

export const CabinAndPassenger = ({
  personCount,
  cabinType,
  onPersonCount,
  onCabinType,
}: CabinAndPassengerProps) => {
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const icon = useMemo(() => {
    if (personCount > 3) {
      return (
        <div style={{ display: 'flex', alignItems: 'center' }}>
          {[...Array(3)].map((_, index) => (
            <ManIcon key={index} sx={{ width: 20 }} />
          ))}
          <AddIcon fontSize='small' sx={{ width: 20 }} />
        </div>
      );
    }

    return (
      <div style={{ display: 'flex' }}>
        {[...Array(personCount)].map((_, index) => (
          <ManIcon key={index} />
        ))}
      </div>
    );
  }, [personCount]);

  return (
    <>
      <IconButton
        onClick={handleClick}
        sx={{
          backgroundColor: 'secondary.dark',
          color: 'white',
          borderRadius: 0,
          width: 90,
          border: `1px solid ${'secondary.dark'}`,
          '&:hover': {
            backgroundColor: 'primary.main',
            border: `1px solid ${'common.white'}`,
          },
        }}
      >
        {icon}
        <Typography py={0.3} px={0.6} fontSize={14} position='absolute' top={0} right={0}>
          {personCount}
        </Typography>
      </IconButton>

      {anchorEl ? (
        <CabinSelectorPopover
          anchorEl={anchorEl}
          onClose={handleClose}
          cabinType={cabinType}
          personCount={personCount}
          onCabinType={onCabinType}
          onPersonCount={onPersonCount}
        />
      ) : null}
    </>
  );
};
