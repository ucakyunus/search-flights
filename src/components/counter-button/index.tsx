import { Box, IconButton, IconButtonProps, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import { Remove, Add } from '@mui/icons-material';

import { CounterActions } from "@/constants";

export interface CounterButtonProps {
  count: number;
  min?: number;
  max?: number;
  onChange: (count: number) => void;
}

const IncrementButton = styled(IconButton)<IconButtonProps>(({ theme }) => ({
  color: theme.palette.secondary.dark,
  width: 30,
  height: 30,
  padding: 0,
  borderRadius: 4
}));

export const CounterButton = ({ count, min = 1, max = 10, onChange }: CounterButtonProps) => {
  
  const handleCountChange = (action: CounterActions) => {
    const newCount =
      action === CounterActions.Increment ? count + 1 : count - 1;
    
    if (newCount >= min && newCount <= max) {
      onChange(newCount);
    }
  };
  
  return (
    <Box display="flex">
      <IncrementButton
        onClick={() => handleCountChange(CounterActions.Decrement)}
        sx={{
          bgcolor: count === min ? 'ButtonShadow' : 'secondary.light'
        }}
      >
        <Remove color="inherit" sx={{ fontSize: 20 }} />
      </IncrementButton>
      
      <Typography fontSize={14} px={2} lineHeight={2}>
        {count}
      </Typography>
      
      <IncrementButton
        onClick={() => handleCountChange(CounterActions.Increment)}
        sx={{
          bgcolor: count === max ? 'ButtonShadow' : 'secondary.light'
        }}
      >
        <Add color="inherit" sx={{ fontSize: 20 }} />
      </IncrementButton>
    </Box>
  );
};