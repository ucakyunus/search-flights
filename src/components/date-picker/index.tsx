import { IconButton, IconButtonProps, Typography } from '@mui/material';
import CalendarMonthIcon from '@mui/icons-material/CalendarMonth';
import { styled } from '@mui/material/styles';

const StyledIconButton = styled(IconButton)<IconButtonProps>(({ theme }) => ({
  backgroundColor: theme.palette.secondary.dark,
  color: 'white',
  borderRadius: 0,
  width: 90,
  border: `1px solid ${theme.palette.secondary.dark}`,
  '&:hover': {
    backgroundColor: theme.palette.primary.main,
    border: `1px solid ${theme.palette.common.white}`,
  },
}));

export const DatePicker = () => {
  return (
    <StyledIconButton size={'large'}>
      <Typography fontSize={14} mr={1.5}>
        Tarih
      </Typography>
      <CalendarMonthIcon fontSize='large' />
    </StyledIconButton>
  );
};
