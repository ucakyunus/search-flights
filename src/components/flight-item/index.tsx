import { useMemo } from 'react';
import { Box, Divider, useTheme } from '@mui/material';

import { CabinOption } from '@/components/flight-item/cabin-option';
import { InfoDetail } from '@/components/flight-item/info-detail';
import { Duration } from '@/components/flight-item/duration';
import { DetailItem } from '@/components/flight-item/detail-item';
import { FareCategories } from '@/constants';

import type { Flight } from '@/types/flight';

export interface FlightItemProps {
  index: number;
  data: Flight;
  isPromotionActive: boolean;
  selectedFareOption: string | null;
  setSelectedFareOption: (value: string) => void;
}

export const FlightItem = (props: FlightItemProps) => {
  const { custom } = useTheme();

  const { data, index, isPromotionActive, selectedFareOption, setSelectedFareOption } = props;

  const { fareCategories } = data;

  const businessRadioValue = `${FareCategories.BUSINESS}-${index}`;
  const economyRadioValue = `${FareCategories.ECONOMY}-${index}`;
  const isBusinessSelected = selectedFareOption === businessRadioValue;
  const isEconomySelected = selectedFareOption === economyRadioValue;

  const flightItemDetailItems = useMemo(() => {
    if (!selectedFareOption) return [];

    if (selectedFareOption !== businessRadioValue && selectedFareOption !== economyRadioValue)
      return [];

    const [selectedCabinType] = selectedFareOption.split('-');

    const subcategories =
      selectedCabinType === FareCategories.BUSINESS
        ? fareCategories.BUSINESS.subcategories
        : fareCategories.ECONOMY.subcategories;

    return subcategories.sort((a, b) => a.order - b.order);
  }, [
    businessRadioValue,
    economyRadioValue,
    fareCategories.BUSINESS.subcategories,
    fareCategories.ECONOMY.subcategories,
    selectedFareOption,
  ]);

  const businessEcoFlyPrice = useMemo(
    () =>
      fareCategories.BUSINESS.subcategories.find((category) => {
        return category.order === 1;
      })?.price.amount as number,
    [fareCategories.BUSINESS.subcategories],
  );

  const economyEcoFlyPrice = useMemo(
    () =>
      fareCategories.ECONOMY.subcategories.find((category) => {
        return category.order === 1;
      })?.price.amount as number,
    [fareCategories.ECONOMY.subcategories],
  );

  const currency = useMemo(
    () => fareCategories.BUSINESS.subcategories[0].price.currency,
    [fareCategories.BUSINESS.subcategories],
  );

  const economyPriceShown = useMemo(() => {
    return isPromotionActive ? economyEcoFlyPrice / 2 : economyEcoFlyPrice;
  }, [economyEcoFlyPrice, isPromotionActive]);

  const businessPriceShown = useMemo(() => {
    return isPromotionActive ? businessEcoFlyPrice / 2 : businessEcoFlyPrice;
  }, [businessEcoFlyPrice, isPromotionActive]);

  return (
    <Box>
      <Box display='flex' gap={2}>
        <Box
          p={2}
          flex={1}
          display='flex'
          alignItems='center'
          justifyContent='space-between'
          boxShadow={custom?.shadows![1]}
          sx={{ backgroundColor: 'white' }}
        >
          <Box display='flex' alignItems='center' flex={1}>
            <InfoDetail
              time={data.departureDateTimeDisplay}
              portCode={data.originAirport.code}
              cityName={data.originAirport.city.name}
            />

            <Divider
              sx={{
                width: '400px',
                height: '1px',
                backgroundColor: '#000',
                mx: 1,
              }}
            />

            <InfoDetail
              textAlign='right'
              time={data.arrivalDateTimeDisplay}
              portCode={data.destinationAirport.code}
              cityName={data.destinationAirport.city.name}
            />
          </Box>

          <Duration mr={3} time={data.flightDuration} />
        </Box>

        <CabinOption
          label='Economy'
          currency={currency}
          price={economyPriceShown}
          isSelected={isEconomySelected}
          onClick={() => setSelectedFareOption(economyRadioValue)}
          radioProps={{
            value: economyRadioValue,
            checked: isEconomySelected,
            onChange: () => setSelectedFareOption(economyRadioValue),
          }}
        />

        <CabinOption
          label='Business'
          currency={currency}
          price={businessPriceShown}
          isSelected={isBusinessSelected}
          onClick={() => setSelectedFareOption(businessRadioValue)}
          radioProps={{
            value: businessRadioValue,
            checked: isBusinessSelected,
            onChange: () => setSelectedFareOption(businessRadioValue),
          }}
        />
      </Box>

      {flightItemDetailItems.length > 0 ? (
        <Box display='flex' gap={2} p={2} mt={2} bgcolor={'white'}>
          {flightItemDetailItems.map((item, index) => (
            <DetailItem key={index} detail={item} isPromotionActive={isPromotionActive} />
          ))}
        </Box>
      ) : null}
    </Box>
  );
};
