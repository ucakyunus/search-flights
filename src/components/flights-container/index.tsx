import React from 'react';
import { Box } from '@mui/material';

export interface FlightsContainerProps {
  children: React.ReactNode;
}

export const FlightsContainer = ({ children }: FlightsContainerProps) => {
  return (
    <Box
      p={2}
      gap={2}
      border={1}
      display='flex'
      flexDirection='column'
      borderColor='secondary.light'
      sx={{ backgroundColor: '#f9f9f9' }}
    >
      {children}
    </Box>
  );
};
