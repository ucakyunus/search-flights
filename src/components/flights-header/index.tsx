import { Box, Button, Typography } from '@mui/material';
import { ListSortBy, ListSortTypes } from '@/constants';
import ArrowUpwardIcon from '@mui/icons-material/ArrowUpward';
import ArrowDownwardIcon from '@mui/icons-material/ArrowDownward';

export interface HeaderProps {
  onSortChange: (type: ListSortBy) => void;
  sortType: ListSortTypes;
}

export const FlightsHeader = ({ onSortChange, sortType }: HeaderProps) => {
  return (
    <Box
      p={1}
      display='flex'
      alignItems='center'
      justifyContent='flex-end'
      sx={{ backgroundColor: 'secondary.dark' }}
    >
      <Typography color='white' mr={2}>
        Sıralama Kriteri
      </Typography>
      <Button
        sx={{
          border: '1px solid #fff',
          mr: 2,
          borderRadius: '4px',
          backgroundColor: 'primary.dark',
          textTransform: 'none',
        }}
        variant='contained'
        onClick={() => onSortChange(ListSortBy.Price)}
      >
        Ekonomi Ücreti
        {sortType === ListSortTypes.PriceAscending && <ArrowUpwardIcon />}
        {sortType === ListSortTypes.PriceDescending && <ArrowDownwardIcon />}
      </Button>
      <Button
        sx={{
          border: '1px solid #fff',
          borderRadius: '4px',
          backgroundColor: 'primary.dark',
          textTransform: 'none',
        }}
        variant='contained'
        onClick={() => onSortChange(ListSortBy.DepartureTime)}
      >
        Kalkış Saati
        {sortType === ListSortTypes.DepartureTimeAscending && <ArrowUpwardIcon />}
        {sortType === ListSortTypes.DepartureTimeDescending && <ArrowDownwardIcon />}
      </Button>
    </Box>
  );
};
