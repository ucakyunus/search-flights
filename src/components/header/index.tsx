import styled from '@emotion/styled';
import { Box, Typography, TypographyProps } from '@mui/material';

export interface HeaderProps {
  bgColor?: 'dark' | 'light';
}

export const Header = ({ bgColor }: HeaderProps) => {
  const dynamicColor = bgColor === 'dark' ? 'white' : 'black';

  const StyledBoldTypography = styled(Typography)(() => ({
    fontSize: '1rem',
    fontWeight: '500',
    color: dynamicColor,
  }));

  const StyledTypography = styled(Typography)<TypographyProps>(() => ({
    fontSize: '1rem',
    fontWeight: '300',
    color: dynamicColor,
  }));

  return (
    <Box
      mt={2}
      mx={2}
      height={16}
      display='flex'
      color='white'
      borderBottom={1}
      alignItems='center'
      borderColor={dynamicColor}
      justifyContent='space-between'
    >
      <StyledBoldTypography>turkishairlines.com</StyledBoldTypography>
      <Box component='span' display='flex'>
        <StyledTypography>search</StyledTypography>
        <StyledBoldTypography>Flight Challange</StyledBoldTypography>
      </Box>
    </Box>
  );
};
