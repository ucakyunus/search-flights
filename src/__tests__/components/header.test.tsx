import { render } from '@testing-library/react';
import { Header, HeaderProps } from '@/components/header';

const mockProps: HeaderProps = {
  bgColor: 'dark',
};

test('Header matches snapshot', () => {
  const { asFragment } = render(<Header {...mockProps} />);
  expect(asFragment()).toMatchSnapshot();
});
