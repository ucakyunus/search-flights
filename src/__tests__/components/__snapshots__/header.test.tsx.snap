// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-z8q3jg"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-nsiw5t-MuiTypography-root"
    >
      turkishairlines.com
    </p>
    <span
      class="MuiBox-root css-k008qs"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-jmii74-MuiTypography-root"
      >
        search
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-nsiw5t-MuiTypography-root"
      >
        Flight Challange
      </p>
    </span>
  </div>
</DocumentFragment>
`;
